analyzers/tests/SonarAnalyzer.Test/Helpers/MessageTemplatesTest.cs

```diff
@@ -0,0 +1,140 @@
+    [DataRow("{")]
+    [DataRow("{{}}")]
+    [DataRow("{{")]
+    [DataRow("}}")]
+    [DataRow("}}{{{{")]
+    [DataRow("hello world")]
+    [DataRow("hello {{world}}")]
+    [DataRow("{{hello {{world}}}}")]
+    public void Parse_NoPlaceholder(string template)
+    {
+        var result = MessageTemplates.Parse(template);
+        ShouldBeSuccess(result);
+    }
+
+    [DataTestMethod]
+    // named
+    [DataRow("{world}", "world", 1, 5)]
+    [DataRow("hello {world}", "world", 7, 5)]
+    [DataRow("hello {{{world42}}}", "world42", 9, 7)]
+    [DataRow("hello {{ {42world}", "42world", 10, 7)]
+    [DataRow("hello {{ {w_0_rld}", "w_0_rld", 10, 7)]
+    [DataRow("hello {{ {_}", "_", 10, 1)]
+    [DataRow("hello {{ {_world}", "_world", 10, 6)]
+    // index
+    [DataRow("{0}", "0", 1, 1)]
+    [DataRow("hello {0}", "0", 7, 1)]
+    [DataRow("hello {{{42}}}", "42", 9, 2)]
+    [DataRow("hello {{ {199}", "199", 10, 3)]
+    // prefix optional operator
+    [DataRow("{@world}", "world", 2, 5)]
+    [DataRow("{$199}", "199", 2, 3)]
+    [DataRow("{@0}", "0", 2, 1)]
+    [DataRow("{$world_42}", "world_42", 2, 8)]
+    public void Parse_Placeholder(string template, string placeholder, int start, int end)
+    {
+        var result = MessageTemplates.Parse(template);
+        ShouldBeSuccess(result, 1);
+        ShouldBe(result.Placeholders[0], placeholder, start, end);
+    }
+
+    [DataTestMethod]
+    // alignment
+    [DataRow("hello {world,1}")]
+    [DataRow("hello {world,42}")]
+    [DataRow("hello {world,-1}")]
+    [DataRow("hello {world,-199}")]
+    // format
+    [DataRow("hello {world:format}")]
+    [DataRow("hello {world:42}")]
+    [DataRow("hello {world:{}")]
+    [DataRow("hello {world:!@#$%^&*()_}")]
+    [DataRow("hello {world:42}")]
+    [DataRow("hello {world:#for_mat42}")]
+    // mixed
+    [DataRow("hello {world,1:format}")]
+    [DataRow("hello {world,42:!@#$%}")]
+    [DataRow("hello {world,-1:dd-MM-yyy}")]
+    [DataRow("hello {world,-42:3,14159}")]
+    public void Parse_Placeholder_Named_Alignment_Format(string template)
+    {
+        var result = MessageTemplates.Parse(template);
+        ShouldBeSuccess(result, 1);
+        ShouldBe(result.Placeholders[0], "world", 7, 5);
+    }
+
+    [TestMethod]
+    public void Parse_Placeholder_Multiple()
+    {
+        var template = "{{what}} {$an_amazing,20} {@day:dd-MM-yyyy} to be alive! {42,-2:format}";
+        var result = MessageTemplates.Parse(template);
+        ShouldBeSuccess(result, 3);
+        ShouldBe(result.Placeholders[0], "an_amazing", 11, 10);
+        ShouldBe(result.Placeholders[1], "day", 28, 3);
+        ShouldBe(result.Placeholders[2], "42", 58, 2);
+    }
+
+    [DataTestMethod]
+
+    //[DataRow("Login failed for {User")]         // Missing closing bracket
+    //[DataRow("Login failed for {{User}")]       //  Opening bracket is escaped
+    //[DataRow("Login failed for {User_Name}")]   //  Only alphanumerics and '_' are allowed for placeholders
```

Comment:
I left these here only for the first round of review, to get an opinion.
They should not be merged.

These 3 cases are from the RSPEC, and whoever specified them thinks they are "invalid".

### About the first two
Semantically, that makes sense, based on the explanation, which shows that it's a "human error".
The problem is, that based on the EBNF of templating, they are syntactically perfectly valid.
Do we care about this? It would only impact [this rule](https://github.com/SonarSource/rspec/pull/2574/files), the rest should be unaffected...And there we need custom handling anyways.

### About the last one

This seems to be correct, both semantically and syntactically. Based on the spec, `_` is not allowed, but [message templates](https://messagetemplates.org/) is perfectly happy with it:

![image](https://github.com/SonarSource/sonar-dotnet/assets/115458417/574816c4-289a-498a-bca6-4fe44a8052c0)

I think we should remove this case from the rspec PR.

Reply:
@gregory-paidis-sonarsource your comment makes sense. Could you please discuss with the people involved in the rule spec to see if there is something particular about why they wanted to have these as invalid cases?

Reply:
It's also worth checking whether the different frameworks allow the use of the `_` character in names.

Reply:
I had a typo and the first one is actually failing as expected.
The second one is valid syntactically, but is probably a typo semantically.
The last one is definitely valid, I will make a small PR on the rspec repo.

@zsolt-kolbay-sonarsource , the frameworks are implementing the message template SPEC, so they should be supporting it: 

![image](https://github.com/SonarSource/sonar-dotnet/assets/115458417/73899620-3994-4993-aecc-fbed08c80748)